### PR TITLE
JDK-8289748: C2 compiled code crashes with SIGFPE with -XX:+StressLCM and -XX:+StressGCM

### PR DESCRIPTION
**Pull request by tobiasholenstein**

### CODE REVIEW

File: src/hotspot/share/opto/loopTransform.cpp [hunk 1]
```diff
@@ -3694,6 +3694,12 @@ bool IdealLoopTree::do_remove_empty_loop(PhaseIdealLoop *phase) {
   Node* exact_limit = phase->exact_limit(this);
   Node* final_iv = new SubINode(exact_limit, cl->stride());
   phase->register_new_node(final_iv, cl->in(LoopNode::EntryControl));
+
+  // There might have been a condition guarding the counted loop.
+  // We need to pin the exact limit to prevent if to float above that check
+  Node* cast_ii = ConstraintCastNode::make(cl->in(LoopNode::EntryControl), exact_limit, phase->_igvn.type(exact_limit), ConstraintCastNode::UnconditionalDependency, T_INT);
+  phase->register_new_node(cast_ii, cl->in(LoopNode::EntryControl));
+  phase->_igvn.replace_input_of(final_iv, 1, cast_ii);
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopTransform.cpp:3702]: Why replace input here instead of using it as input when `final_iv` is created?

Comment (tobiasholenstein) [reply src/hotspot/share/opto/loopTransform.cpp]: Hi Vladimir,
You are right. I updated to directly create with `cast_ii` as input to `final_iv`

File: src/hotspot/share/opto/loopTransform.cpp [hunk 2]
```diff
@@ -3692,7 +3692,13 @@ bool IdealLoopTree::do_remove_empty_loop(PhaseIdealLoop *phase) {
   // counted loop has limit check predicate.
   Node* phi = cl->phi();
   Node* exact_limit = phase->exact_limit(this);
-  Node* final_iv = new SubINode(exact_limit, cl->stride());
+
+  // There might have been a condition guarding the counted loop.
+  // We need to pin the exact limit to prevent if to float above that check
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopTransform.cpp:3697]:
```suggestion
  // We need to pin the exact limit to prevent it from floating above that check
```

**Review by TobiHartmann — APPROVED** [review]
Nice analysis! Looks good to me.

File: src/hotspot/share/opto/loopTransform.cpp [hunk 3]
```diff
@@ -3692,7 +3692,13 @@ bool IdealLoopTree::do_remove_empty_loop(PhaseIdealLoop *phase) {
   // counted loop has limit check predicate.
   Node* phi = cl->phi();
   Node* exact_limit = phase->exact_limit(this);
-  Node* final_iv = new SubINode(exact_limit, cl->stride());
+
+  // There might have been a condition guarding the counted loop.
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopTransform.cpp:3696]: There should always be a zero trip guard (ensured by code above). So, I would rephrase this to make it more clear that we always need to do the pin on this guard (or remove this phrase completely).

**Review by chhagedorn — APPROVED** [review]
Looks good!

File: src/hotspot/share/opto/loopTransform.cpp (hunk 3 above)
Comment (tobiasholenstein) [reply src/hotspot/share/opto/loopTransform.cpp]: Thx. I rephrased the comment

**Review by TobiHartmann — APPROVED** [review]

**Comment by tobiasholenstein** [comment]
@vnkozlov , @TobiHartmann and @chhagedorn thanks for the reviews!

/integrate
